enterprise/reporting/error-index/worker.go

```diff
@@ -0,0 +1,386 @@
+		w.config.concurrency.Load(),
+		w.statsFactory,
+	)
+	w.limiter.update = kitsync.NewLimiter(
+		w.lifecycle.ctx, &w.limiterGroup, "erridx_update",
+		w.config.concurrency.Load(),
+		w.statsFactory,
+	)
+	return w
+}
+
+func (w *worker) Work() (didSomeWork bool) {
+	for {
+		select {
+		case <-w.lifecycle.ctx.Done():
+			return
+		default:
+			didWork, didLimitReached := w.doWork()
+			didSomeWork = didSomeWork || didWork // Accumulate work status
+
+			// If no work was done or limit not reached, exit the loop
+			if !didWork || !didLimitReached {
+				return
+			}
+		}
+	}
+}
+
+// doWork does the following:
+// 1. Fetches job results. If no jobs are fetched, returns.
+// 2. Checks if job limits are not reached and upload frequency is not met; returns.
+// 3. Upload jobs to object storage.
+// 4. Updates job status in jobsDB.
+func (w *worker) doWork() (didWork, didLimitReached bool) {
+	jobResult, err := w.fetchJobs()
+	if err != nil && w.lifecycle.ctx.Err() != nil {
+		return
+	}
+	if err != nil {
+		panic(fmt.Errorf("failed to fetch jobs for error index: %s", err.Error()))
+	}
+	if len(jobResult.Jobs) == 0 {
+		return
+	}
+	if !jobResult.LimitsReached && time.Since(w.lastUploadTime) < w.config.uploadFrequency {
+		return
+	}
+
+	statusList, err := w.uploadJobs(w.lifecycle.ctx, jobResult.Jobs)
+	if err != nil {
+		w.log.Errorw("failed to upload jobs", "error", err)
+		return
+	}
+	w.lastUploadTime = w.now()
+
+	err = w.markJobsStatus(statusList)
+	if err != nil && w.lifecycle.ctx.Err() != nil {
+		return
+	}
+	if err != nil {
+		panic(fmt.Errorf("failed to mark jobs: %s", err.Error()))
+	}
+
+	tags := stats.Tags{
+		"workspaceId": w.workspaceID,
+		"sourceId":    w.sourceID,
+	}
+	w.statsFactory.NewTaggedStat("erridx_uploaded_jobs", stats.CountType, tags).Count(len(jobResult.Jobs))
+
+	return true, jobResult.LimitsReached
+}
+
+func (w *worker) fetchJobs() (jobsdb.JobsResult, error) {
+	defer w.limiter.fetch.Begin(w.sourceID)()
+
+	return w.jobsDB.GetUnprocessed(w.lifecycle.ctx, jobsdb.GetQueryParams{
+		ParameterFilters: []jobsdb.ParameterFilterT{
+			{Name: "source_id", Value: w.sourceID},
+		},
+		PayloadSizeLimit: w.config.payloadLimit.Load(),
+		EventsLimit:      int(w.config.eventsLimit.Load()),
+		JobsLimit:        int(w.config.eventsLimit.Load()),
+	})
+}
+
+// uploadJobs uploads aggregated job payloads to object storage.
+// It aggregates payloads from a list of jobs, applies transformations if needed,
+// uploads the payloads, and returns the concatenated locations of the uploaded files.
+func (w *worker) uploadJobs(ctx context.Context, jobs []*jobsdb.JobT) ([]*jobsdb.JobStatusT, error) {
+	defer w.limiter.upload.Begin(w.sourceID)()
+
+	jobPayloadsMap, err := aggregatedJobPayloads(jobs)
+	if err != nil {
+		return nil, fmt.Errorf("aggregating job payloads: %w", err)
+	}
+
+	if w.configFetcher.IsPIIReportingDisabled(w.workspaceID) {
+		for _, jobPayloads := range jobPayloadsMap {
+			transformPayloadsForPII(jobPayloads)
+		}
+	}
```

Comment:
I'm not sure we need to do this, yet

Reply:
I thought, since we are doing it for error reporting, we might need to do this at some stage. @lvrach @cisse21 wdyt?

Reply:
Based on the [discussion](https://rudderlabs.slack.com/archives/C01HTT66UMB/p1698223718473169) in here, removing pii transformations.